include/swift/SILOptimizer/Analysis/FunctionSignatureAnalysis.h

```diff
@@ -0,0 +1,265 @@
+
+  /// If non-null, this is the release in the return block of the callee, which
+  /// is associated with this parameter if it is @owned. If the parameter is not
+  /// @owned or we could not find such a release in the callee, this is null.
+  RetainList CalleeRetain;
+
+  /// Initialize this argument descriptor with all information from A that we
+  /// use in our optimization.
+  ///
+  /// *NOTE* We cache a lot of data from the argument and maintain a reference
+  /// to the original argument. The reason why we do this is to make sure we
+  /// have access to the original argument's state if we modify the argument
+  /// when optimizing.
+  ResultDescriptor() {};
+  ResultDescriptor(SILResultInfo RI) : ResultInfo(RI), CalleeRetain() {}
+
+  ResultDescriptor(const ResultDescriptor &) = delete;
+  ResultDescriptor(ResultDescriptor &&) = default;
+  ResultDescriptor &operator=(const ResultDescriptor &) = delete;
+  ResultDescriptor &operator=(ResultDescriptor &&) = default;
+
+  /// \returns true if this argument's ParameterConvention is P.
+  bool hasConvention(ResultConvention R) const {
+    return ResultInfo.getConvention() == R;
+  }
+};
+
+class FunctionSignatureFunctionInfo {
+  /// Function currently analyzing.
+  SILFunction *F;
+
+  /// The alias analysis currently using.
+  AliasAnalysis *AA;
+
+  /// The rc-identity analysis currently using.
+  RCIdentityFunctionInfo *RCFI;
+
+  /// Does any call inside the given function may bind dynamic 'Self' to a
+  /// generic argument of the callee.
+  bool MayBindDynamicSelf;
+
+  /// Did we decide to change the self argument? If so we need to
+  /// change the calling convention 'method' to 'freestanding'.
+  bool ShouldModifySelfArgument = false;
+
+  /// A list of structures which present a "view" of precompiled information on
+  /// an argument that we will use during our optimization.
+  llvm::SmallVector<ArgumentDescriptor, 8> ArgDescList;
+
+  /// Keep a "view" of precompiled information on the direct results
+  /// which we will use during our optimization.
+  llvm::SmallVector<ResultDescriptor, 4> ResultDescList;
+
+  llvm::BumpPtrAllocator Allocator;
+
+public:
+  FunctionSignatureFunctionInfo(SILFunction *F, AliasAnalysis *AA,
+                                RCIdentityFunctionInfo *RCFI) :
+  F(F), AA(AA), RCFI(RCFI), MayBindDynamicSelf(computeMayBindDynamicSelf(F)) {}
+
+  bool analyze();
+  bool analyzeParameters();
+  bool analyzeResult();
+
+  /// Returns the mangled name of the function that should be generated from
+  /// this function analyzer.
+  std::string getOptimizedName() const;
+
+  bool shouldModifySelfArgument() const { return ShouldModifySelfArgument; }
+  ArrayRef<ArgumentDescriptor> getArgDescList() const { return ArgDescList; }
+  MutableArrayRef<ResultDescriptor> getResultDescList() {return ResultDescList;}
+  MutableArrayRef<ArgumentDescriptor> getArgDescList() { return ArgDescList; }
+  SILFunction *getAnalyzedFunction() const { return F; }
```

Comment:
We should really only be vending non-mutable arrays of const values to clients, so that no client can modify the results of the analysis after it has been computed.